InteractionBase/src/vis/ViewSwitcherMenu.cpp

```diff
@@ -0,0 +1,98 @@
+ **
+ **
+ ** THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES,
+ ** INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ ** DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ ** SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ ** SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+ ** WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ ** OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ **
+ **********************************************************************************************************************/
+#include "ViewSwitcherMenu.h"
+
+#include "vis/VViewSwitcherEntry.h"
+#include "VisualizationBase/src/items/ViewItem.h"
+#include "VisualizationBase/src/cursor/TextCursor.h"
+
+namespace Interaction {
+
+ITEM_COMMON_DEFINITIONS(ViewSwitcherMenu, "item")
+
+void ViewSwitcherMenu::show(Visualization::Item* target)
+{
+	QApplication::postEvent(target->scene(),
+							new Visualization::CustomSceneEvent( [=]() { showNow(target); }));
+}
+
+void ViewSwitcherMenu::showNow(Visualization::Item* target)
+{
+	QVector<Visualization::Item*> items;
+	for (auto view : target->scene()->viewItems())
+		items.append(new VViewSwitcherEntry(nullptr, view->name()));
+	for (int i = items.size(); i < 9; i++)
+		items.append(new VViewSwitcherEntry(nullptr, "Empty slot " + QString::number(i)));
+	Menu::hideNow();
+	Menu::instance = new ViewSwitcherMenu(items, target);
+	target->scene()->addTopLevelItem(Menu::instance);
+	target->scene()->addPostEventAction( [=]()
+					{ Menu::instance->selectItem(Menu::instance->currentItems()[0][0]); });
+}
+
+ViewSwitcherMenu::ViewSwitcherMenu(QVector<Visualization::Item*> items,
+											 Visualization::Item* target, StyleType* style)
+	: Super(items, target, style)
+{
+}
+
+void ViewSwitcherMenu::startFocusMode(Visualization::Item *target)
+{
+	if (auto asSwitcher = DCast<VViewSwitcherEntry>(target))
+	{
+		nameBefore_ = asSwitcher->nameField()->text();
+		asSwitcher->setGraphicsEffect(nullptr);
+		asSwitcher->nameField()->moveCursor();
+		asSwitcher->nameField()->correspondingSceneCursor<Visualization::TextCursor>()->selectAll();
+	}
+}
+
+void ViewSwitcherMenu::endFocusMode(Visualization::Item *target)
+{
+	if (auto asSwitcher = DCast<VViewSwitcherEntry>(target))
+	{
+		//If the view item didn't exist and the user changed the name,
+		//create a new item with that name
+		auto nameAfter = asSwitcher->nameField()->text();
+		if (scene()->viewItem(nameAfter) == nullptr && nameAfter != nameBefore_)
+			scene()->newViewItem(nameAfter);
+	}
+}
+
+bool ViewSwitcherMenu::onSelectItem(Visualization::Item* item)
```

Comment:
This method should really be called `onExecute`, since selecting is not the same as selecting + executing the current selection. Actually, the shorter name `execute` is even better if you ask me. `executeEntry` might also work.